Fixed #29789 -- Added support for nested relations to FilteredRelation.

ferrants opened the pull request:
For ticket 29789
https://code.djangoproject.com/ticket/29789

felixxm commented on tests/filtered_relation/tests.py:
```diff
@@ -156,8 +156,15 @@ def test_with_multiple_filter(self):
+                book_title_alice__isnull=False
+            ).filter(
+                book_title_alice__isnull=False
+            ).distinct(),
```
This change is unrelated, please revert it.

felixxm commented on tests/filtered_relation/tests.py:
```diff
@@ -248,10 +255,11 @@ def test_difference(self):
-            Author.objects.annotate(
-                book_jane=FilteredRelation('book', condition=Q(book__title__iexact='the book by jane a')),
-            ).filter(book_jane__isnull=False).select_for_update(),
+            qs,
```
This change is unrelated, please revert it.

ferrants replied:
addressed

felixxm commented on tests/filtered_relation/tests.py:
```diff
@@ -279,28 +287,153 @@ def test_as_subquery(self):
+                'book', condition=Q(book__title__icontains='book')
+            )
+        )
+        qs = qs.annotate(
```
I would prefer to chain `annotate()`, e.g.
```python
...
).annotate(
...
```

ferrants replied:
addressed!

felixxm commented on django/db/models/sql/query.py:
```diff
@@ -1408,13 +1408,23 @@ def build_filtered_relation_q(self, q_object, reuse, branch_negated=False, curre
+            for _ in relation_field_parts:
+                if lookup_path:
+                    lookup_path = lookup_path[1:]
+            if lookup_path:
```
The above five lines can be simplified to:
```python
    ...
    shift = 2 if not lookup_parts else 1
    if len(field_parts) > (shift + len(relation_field_parts)):
    raise ValueError(...
```
so we don't need a `lookup_path` variable and iterate over the `relation_field_parts`.

ferrants replied:
addressed!

ferrants commented:
@felixxm , i addressed your feedback and have rebased. any other feedback is greatly appreciated!

felixxm commented on django/db/models/sql/query.py:
```diff
@@ -1408,13 +1408,19 @@ def build_filtered_relation_q(self, q_object, reuse, branch_negated=False, curre
                 raise ValueError(
                     "FilteredRelation's condition doesn't support nested "
-                    "relations (got %r)." % lookup
+                    "on clauses deeper than the relation (got %r for %r)." % (lookup, filtered_relation.relation_name)
```
Validation based only on a nesting level is not a proper solution, e.g.
```python
Author.objects.annotate(
    book_editor_worked_with=FilteredRelation('book__editor', condition=Q(book__author__name__icontains='book')),
)
``` 
will crash with `django.db.utils.OperationalError: no such column: T4.name`.

ferrants replied:
I'll add a test case specifically for this, thanks

ferrants replied:
Added test to catch this error

felixxm commented on tests/filtered_relation/tests.py:
```diff
@@ -279,28 +279,149 @@ def test_as_subquery(self):
+                    'book__title__icontains',
+                    condition=Q(book__editor__name__icontains='b')
+                ),
+            ).filter(book_edited_by_b__isnull=False).distinct()
```
`distinct()` is not necessary for this test, please remove it.

felixxm commented on tests/filtered_relation/tests.py:
```diff
@@ -279,28 +279,149 @@ def test_as_subquery(self):
-            ))
+            qs = Author.objects.annotate(
+                book_edited_by_b=FilteredRelation('book', condition=Q(book__editor__name__icontains='b')),
+            ).filter(book_edited_by_b__isnull=False).distinct()
```
`distinct()` is not necessary for this test, please remove it.

felixxm commented on tests/filtered_relation/tests.py:
```diff
@@ -279,28 +279,149 @@ def test_as_subquery(self):
+
+    def test_with_nested_field(self):
+        qs = Author.objects.annotate(
+            book_editor_worked_with=FilteredRelation('book__editor', condition=Q(book__title__icontains='book')),
```
I'm not sure if that is a correct example and sth that we would like to support :thinking:, because conditions across different relations introduce duplicate rows. IMO we should restrict conditions only to starting with `filtered_relation`.

ferrants replied:
What about restricting to those within the `filtered_relation` path? `book__title__icontains` comes from the `book` table, which is accessed on the way to `book_editor`

For more context, I'm using this to do a join on the right side of a left join to convert currencies. I could get more into my specific models if that'd be helpful, but don't want them to distract from the basic functionality. I didn't want to add to the models under test here to replicate what I'm doing, but maybe I should to make it more explicit.

ferrants replied:
@felixxm , I added restrictions around this so the `filtered_relation`'s condition must be within the path to the `filtered_relation`

ferrants commented on django/db/models/sql/query.py:
```diff
@@ -1408,13 +1408,41 @@ def build_filtered_relation_q(self, q_object, reuse, branch_negated=False, curre
-            if len(field_parts) > (shift + len(lookup_parts)):
+            field_path = field_parts[:-shift]
+
+            if len(field_path) <= len(relation_field_parts):
```
I reverted this back to looping through the relationship and lookup fields so I can have better checks to make sure the lookup is in the relationship path. There is probably more I should add to this. Thinking about it.

ferrants commented on django/db/models/sql/query.py:
```diff
@@ -1408,13 +1408,41 @@ def build_filtered_relation_q(self, q_object, reuse, branch_negated=False, curre
+                            lookup, filtered_relation.relation_name)
+                        )
+            else:
                 raise ValueError(
```
Probably not the right error case, it's also duplicate of the one right above it, checking inside the nested one. This can probably be removed, but I am considering if there are other errors or checks that should be made.

ferrants commented:
@felixxm , rebased and updated the FilteredRelation condition check to be more strict, enforcing it to be within the path to the FilteredRelation, along with some more tests. Could you take another look? Thanks!

ferrants commented:
@felixxm , I've rebased this, can you take another look? Thanks!

ferrants commented:
@felixxm , I've rebased this and updated the ticket, can someone take another look? Is there another way I should be reaching out about this? Thanks!

auvipy commented:
you should assign the ticket to you comment there. I think it's already in the review Q

ferrants commented:
@auvipy , good call, thanks. Assigned to me

ferrants commented:
@felixxm , I've rebased this again, any chance I could get another review? Thanks!

ferrants commented on django/db/models/sql/query.py:
```diff
@@ -1407,14 +1407,32 @@ def build_filtered_relation_q(self, q_object, reuse, branch_negated=False, curre
+                if len(relation_field_parts) > idx:
+                    if relation_field_parts[idx] != lookup_field_part:
+                        raise ValueError(
+                            "FilteredRelation's condition doesn't support nested "
```
Good call, thanks

ferrants commented:
Rebased! Still good for review and merge!

ferrants commented:
Rebased, still good to go! Been using this patch for months and it does the trick. Nested FilteredRelations are pretty useful for analytical queries.

adamchainz commented on tests/filtered_relation/tests.py:
```diff
@@ -279,28 +279,184 @@ def test_as_subquery(self):
+                    'book__title__icontains',
+                    condition=Q(book__editor__name__icontains='b')
+                ),
+            ).filter(book_edited_by_b__isnull=False)
+            list(qs)
```
Afaics this message should be triggered in the `.annotate()`, so the `filter()` and `list()` aren't needed?

adamchainz commented on tests/filtered_relation/tests.py:
```diff
@@ -279,28 +279,184 @@ def test_as_subquery(self):
-            ))
+            qs = Author.objects.annotate(
+                book_edited_by_b=FilteredRelation('book', condition=Q(book__editor__name__icontains='b')),
+            ).filter(book_edited_by_b__isnull=False)
+            list(qs)
```
Afaics this message should be triggered in the `.annotate()`, so the `filter()` and `list()` aren't needed?

ferrants commented:
Thanks @adamchainz ! I addressed your comments and added to the 3.2 release notes

ferrants commented on tests/filtered_relation/tests.py:
```diff
@@ -279,28 +286,182 @@ def test_as_subquery(self):
+                ),
+            ).filter(book_edited_by_b__isnull=False)
+
+    def test_with_foreign_key_relation_name_lookup_not_within_path(self):
```
@charettes , this test asserts that the "right side" of the condition can't be deeper, which should avoid to N-to-Many relationship nested query issue.

ferrants commented on tests/filtered_relation/tests.py:
```diff
@@ -279,28 +286,182 @@ def test_as_subquery(self):
+                (self.book4, self.editor_b),
+            ], lambda x: (x, x.author_favorite_book_editor))
+
+    def test_with_foreign_key_on_condition_deeper_than_relationship_error(self):
```
@charettes , I've considered your comment and I think that the current limitations around lookup usages prevent that from happening. This test asserts that the "left side" of the condition can't be deeper, which should avoid to N-to-Many relationship nested query issue.

ferrants commented on tests/filtered_relation/tests.py:
```diff
@@ -424,3 +585,182 @@ def test_aggregate(self):
+            seller=cls.seller2,
+        )
+
+    def test_with_summation_aggregation(self):
```
This test reflects how I am actually using this feature to get more performance out of some queries by improving the join-on logic.

ferrants commented on tests/filtered_relation/tests.py:
```diff
@@ -424,3 +585,182 @@ def test_aggregate(self):
+        qs = Book.objects.annotate(
+            recent_sales=FilteredRelation(
+                'daily_sales',
+                condition=~Q(daily_sales__seller=self.seller1)
```
@charettes , this case is most similar to what you suggested we have tests cover, using negation on a nested relationship.

charettes replied:
This code path won't trigger `split_exclude` but from my local testing it seems `condition=~Q` already has a flawed implementation on `master` anyway.

```diff
     def test_with_exclude(self):
+        self.assertSequenceEqual(
+            Author.objects.annotate(
+                book_alice=FilteredRelation('book', condition=Q(book__title__iexact='poem by alice')),
+            ).filter(book_alice__isnull=True),
+            [self.author2],
+        )
         self.assertSequenceEqual(
             Author.objects.annotate(
                 book_alice=FilteredRelation('book', condition=Q(book__title__iexact='poem by alice')),
             ).exclude(book_alice__isnull=False),
             [self.author2],
         )
+        self.assertSequenceEqual(
+            Author.objects.annotate(
+                book_alice=FilteredRelation('book', condition=Q(book__title__iexact='poem by alice')),
+            ).filter(book_alice__isnull=False),
+            [self.author1],
+        )
+        self.assertSequenceEqual(
+            Author.objects.annotate(
+                book_alice=FilteredRelation('book', condition=Q(book__title__iexact='poem by alice')),
+            ).exclude(book_alice__isnull=True),
+            [self.author1],
+        )
+        self.assertSequenceEqual(
+            Author.objects.annotate(
+                book_not_alice=FilteredRelation('book', condition=~Q(book__title__iexact='poem by alice')),
+            ).filter(book_not_alice__isnull=False),
+            [self.author2, self.author2, self.author1],
+        )
+        # The following fails while all the other succeed.
+        self.assertSequenceEqual(
+            Author.objects.annotate(
+                book_not_alice=FilteredRelation('book', condition=~Q(book__title__iexact='poem by alice')),
+            ).exclude(book_not_alice__isnull=True),
+            [self.author2, self.author2, self.author1],
+        )

     def test_with_join_and_complex_condition(self):
         self.assertSequenceEqual(
```

charettes replied:
To summarize the issue mentioned above is not related to the patch here, looks like it already exists in master.

ferrants replied:
Thanks for the `--debug-sql` tip, I wasn't aware of that. And thanks for those examples of what you mean. It looks like there was a failure in the `database=oracle18,label=bionic4,python=python3.8` tests and now after rebasing in the spelling checks in the docs, so I'll look into those tests today as well to see if I can get a better assertion in place.

ferrants replied:
@charettes , Some other tests in this suite use a `distinct` to I think avoid the same issue, would you feel comfortable using a distinct to assert the right result? For example, in the last example that you posted, if you add a distinct, this passes:
```
        self.assertSequenceEqual(
            Author.objects.annotate(
                book_not_alice=FilteredRelation('book', condition=~Q(book__title__iexact='poem by alice')),
            ).filter(book_not_alice__isnull=False).order_by('id').distinct(),
            [self.author1, self.author2],
        )
```
with query:
```
SELECT DISTINCT "filtered_relation_author"."id", "filtered_relation_author"."name", "filtered_relation_author"."content_type_id", "filtered_relation_author"."object_id" FROM "filtered_relation_author" INNER JOIN "filtered_relation_book" book_not_alice ON ("filtered_relation_author"."id" = book_not_alice."author_id" AND (NOT (UPPER(book_not_alice."title"::text) = UPPER(poem by alice)))) WHERE book_not_alice."id" IS NOT NULL ORDER BY "filtered_relation_author"."id" ASC
```

I don't think it make sense to assert the `[self.author1, self.author2, self.author2]` result for the non-distinct one just to document the current behavior, we should probably try to push-down to a sub-query in that case, right? Do we have a separate ticket for the existing issue in `master`? There are some closed-fixed ones here that look related: https://code.djangoproject.com/search?q=split_exclude&noquickjump=1&changeset=on&milestone=on&ticket=on&wiki=on

charettes replied:
@ferrants I think this can be addressed in a separate ticket yes, the above diff was more a of demonstration of current issues with negated filtered conditions but you're probably right that it's related to existing tickets.

felixxm commented:
buildbot, test on oracle.

felixxm commented:
@ferrants I cannot push my final edits :disappointed: Can you check 
![obraz](https://user-images.githubusercontent.com/2865885/87023117-a2b32400-c1d7-11ea-8190-cb20cf8dc13f.png)?

felixxm commented:
I pushed final edits to a new PR #13170
